clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonInstanceTypeCachingAgent.java

```diff
@@ -0,0 +1,196 @@
+          EntityUtils.consumeQuietly(getResponse.getEntity());
+          throw new Exception("failed to read instance type data for " + region + ": " + getResponse.getStatusLine().toString());
+        }
+        Optional<String> etag = getEtagHeader(getResponse);
+        HttpEntity entity = getResponse.getEntity();
+        instanceTypes = fromStream(entity.getContent());
+        EntityUtils.consumeQuietly(entity);
+        if (etag.isPresent()) {
+          metadataAttributes = new MetadataAttributes();
+          metadataAttributes.etag = etag.get();
+          metadataAttributes.cachedInstanceTypes = new HashSet<>(instanceTypes);
+          metadata = new DefaultCacheData("metadata", objectMapper.convertValue(metadataAttributes, ATTRIBUTES), Collections.emptyMap());
+        } else {
+          metadata = null;
+        }
+      }
+      Map<String, Collection<String>> evictions = new HashMap<>();
+      Map<String, Collection<CacheData>> cacheResults = new HashMap<>();
+      List<CacheData> instanceTypeData = new LinkedList<>();
+      cacheResults.put(Keys.Namespace.INSTANCE_TYPES.getNs(), instanceTypeData);
+      if (metadata != null) {
+        cacheResults.put(getAgentType(), Collections.singleton(metadata));
+      } else {
+        evictions.put(getAgentType(), Collections.singleton("metadata"));
+      }
+
+      for (String instanceType : instanceTypes) {
+        for (String account : matchingAccounts) {
+          Map<String, Object> instanceTypeAttributes = new HashMap<>();
+          instanceTypeAttributes.put("account", account);
+          instanceTypeAttributes.put("region", region);
+          instanceTypeAttributes.put("name", instanceType);
+          instanceTypeData.add(new DefaultCacheData(Keys.getInstanceTypeKey(instanceType, region, account), instanceTypeAttributes, Collections.emptyMap()));
+        }
+      }
+
+      return new DefaultCacheResult(cacheResults, evictions);
+    } catch (Exception ex) {
+      throw new RuntimeException(ex);
+    }
+  }
+
+  Optional<String> getEtagHeader(HttpResponse response) {
+    return Optional.ofNullable(response).map(r -> r.getFirstHeader("ETag")).map(Header::getElements).filter(e -> e.length > 0).map(e -> e[0].getName());
+  }
+
+  @Override
+  public String getAgentType() {
+    return getClass().getSimpleName() + "/" + region;
+  }
+
+  @Override
+  public String getProviderName() {
+    return AwsInfrastructureProvider.PROVIDER_NAME;
+  }
+
+  static class Offering {
+    public String productFamily;
+    public ComputeInstanceAttributes attributes;
+  }
+
+  static class ComputeInstanceAttributes {
+    public String instanceType;
+
+    @Override
+    public String toString() {
+      return instanceType;
+    }
+  }
+
+  static class Offerings {
+    public Map<String, Offering> products;
+  }
+
+  static class MetadataAttributes {
+    public String etag;
+    public Set<String> cachedInstanceTypes;
+  }
+
+
+  //visible for testing
```

Comment:
curious re: this vs. the ol `@VisibleForTesting` annotation?

Reply:
not really a fan of importing a guava annotation when nothing else is guava'd and we don't really get much of anything tooling wise from that annotation